test/general/test_entrance_rando.py

```diff
@@ -0,0 +1,376 @@
+import unittest
+from enum import IntEnum
+from typing import List, Type
+
+from BaseClasses import Region, EntranceType, MultiWorld, Entrance
+from entrance_rando import disconnect_entrance_for_randomization, randomize_entrances, EntranceRandomizationError, \
+    ERPlacementState, EntranceLookup, bake_target_group_lookup
+from Options import Accessibility
+from test.general import generate_test_multiworld, generate_locations, generate_items
+from worlds.generic.Rules import set_rule
+
+
+class ERTestGroups(IntEnum):
+    LEFT = 1
+    RIGHT = 2
+    TOP = 3
+    BOTTOM = 4
+
+
+directionally_matched_group_lookup = {
+    ERTestGroups.LEFT: [ERTestGroups.RIGHT],
+    ERTestGroups.RIGHT: [ERTestGroups.LEFT],
+    ERTestGroups.TOP: [ERTestGroups.BOTTOM],
+    ERTestGroups.BOTTOM: [ERTestGroups.TOP]
+}
+
+
+def generate_entrance_pair(region: Region, name_suffix: str, group: int):
+    lx = region.create_exit(region.name + name_suffix)
+    lx.randomization_group = group
+    lx.randomization_type = EntranceType.TWO_WAY
+    le = region.create_er_target(region.name + name_suffix)
+    le.randomization_group = group
+    le.randomization_type = EntranceType.TWO_WAY
+
+
+def generate_disconnected_region_grid(multiworld: MultiWorld, grid_side_length: int, region_size: int = 0,
+                                      region_type: Type[Region] = Region):
+    """
+    Generates a grid-like region structure for ER testing, where menu is connected to the top-left region, and each
+    region "in vanilla" has 2 2-way exits going either down or to the right, until reaching the goal region in the
+    bottom right
+    """
+    for row in range(grid_side_length):
+        for col in range(grid_side_length):
+            index = row * grid_side_length + col
+            name = f"region{index}"
+            region = region_type(name, 1, multiworld)
+            multiworld.regions.append(region)
+            generate_locations(region_size, 1, region=region, tag=f"_{name}")
+
+            if row == 0 and col == 0:
+                multiworld.get_region("Menu", 1).connect(region)
+            if col != 0:
+                generate_entrance_pair(region, "_left", ERTestGroups.LEFT)
+            if col != grid_side_length - 1:
+                generate_entrance_pair(region, "_right", ERTestGroups.RIGHT)
+            if row != 0:
+                generate_entrance_pair(region, "_top", ERTestGroups.TOP)
+            if row != grid_side_length - 1:
+                generate_entrance_pair(region, "_bottom", ERTestGroups.BOTTOM)
+
+
+class TestEntranceLookup(unittest.TestCase):
+    def test_shuffled_targets(self):
+        """tests that get_targets shuffles targets between groups when requested"""
+        multiworld = generate_test_multiworld()
+        generate_disconnected_region_grid(multiworld, 5)
+
+        lookup = EntranceLookup(multiworld.per_slot_randoms[1], coupled=True)
```

Comment:
Deprecated, may be hard-deprecated eventually. Use `multiworld.worlds[1].random`

Reply:
To my credit, at the time these were originally written, some peculiarities of world loading prevented this from working because world.random was not getting initialized in tests, fixing that was considered out of scope for this PR.